src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -1,654 +1,650 @@
 
     /**
-     * Chooses a target from the list.
+     * Get a target for the guard.
      *
-     * @return the next state.
+     * @return The next AIState to go to.
      */
-    protected AIState getTarget()
+    protected EntityLivingBase getTarget()
     {
-        if (entityList.isEmpty())
-        {
-            return GUARD_PATROL;
-        }
-
-        final Entity entity = entityList.get(0);
+        final AbstractBuildingGuards building = (AbstractBuildingGuards) getOwnBuilding();
 
-        final BlockPos buildingLocation = getOwnBuilding().getLocation();
-
-        //Only attack entities in max patrol distance.
-        if (BlockPosUtil.getDistance2D(entity.getPosition(), buildingLocation) < ((AbstractBuildingGuards) getOwnBuilding()).getPatrolDistance())
+        if (building != null && target == null)
         {
-            if (worker.getEntitySenses().canSee(entity) && (entityList.get(0)).isEntityAlive())
+            final List<EntityLivingBase> targets = world.getEntitiesWithinAABB(EntityLivingBase.class, getSearchArea());
+
+            for (final EntityLivingBase entity : targets)
             {
                 if (entity instanceof EntityPlayer)
                 {
-                    if (worker.getColony() != null && worker.getColony().getPermissions().hasPermission((EntityPlayer) entity, Action.GUARDS_ATTACK))
+                    final EntityPlayer player = (EntityPlayer) entity;
+
+                    if (worker.getColony() != null && worker.getColony().getPermissions().hasPermission(player, Action.GUARDS_ATTACK))
                     {
-                        targetEntity = (EntityLivingBase) entity;
-                        worker.getNavigator().clearPath();
-                        return GUARD_HUNT_DOWN_TARGET;
+                        return entity;
                     }
-                    entityList.remove(0);
-                    setDelay(BASE_DELAY);
-                    return GUARD_GET_TARGET;
-                }
-                else
-                {
-                    worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.hunting"));
 
-                    worker.getNavigator().clearPath();
-                    targetEntity = (EntityLivingBase) entity;
-                    return GUARD_HUNT_DOWN_TARGET;
                 }
             }
 
-            if (!(entityList.get(0)).isEntityAlive())
+            float closest = -1;
+            EntityLivingBase targetEntity = null;
+
+            for (final MobEntryView mobEntry : building.getMobsToAttack())
             {
-                return GUARD_GATHERING;
+
+                if (mobEntry.hasAttack())
+                {
+                    for (final EntityLivingBase entity : targets)
+                    {
+                        if (mobEntry.getEntityEntry().getEntityClass().isInstance(entity)
+                              && ( worker.getDistance(entity) < closest
+                              || closest == -1))
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Equality tests should not be made with floating point values. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1244)